runtime/obj.c

```diff
@@ -250,6 +212,74 @@ CAMLprim value caml_lazy_make_forward (value v)
   CAMLreturn (res);
 }
 
+static int obj_update_tag (value blk, int old_tag, int new_tag)
+{
+  header_t hd;
+  tag_t tag;
+
+  SPIN_WAIT {
+    hd = Hd_val(blk);
+    tag = Tag_hd(hd);
+
+    if (tag != old_tag) return 0;
+    if (caml_domain_alone()) {
+      Tag_val (blk) = new_tag;
+      return 1;
+    }
+
+    if (atomic_compare_exchange_strong(Hp_atomic_val(blk), &hd,
+                                       (hd & ~0xFF) | new_tag))
+      return 1;
+  }
+}
+
+CAMLprim value caml_lazy_reset_to_lazy (value v)
+{
+  CAMLassert (Tag_val(v) == Forcing_tag);
+
+  obj_update_tag (v, Forcing_tag, Lazy_tag);
+  return Val_unit;
+}
+
+CAMLprim value caml_lazy_update_to_forward (value v)
+{
+  CAMLassert (Tag_val(v) == Forcing_tag);
+
+  obj_update_tag (v, Forcing_tag, Forward_tag);
+  return Val_unit;
+}
+
+CAMLprim value caml_lazy_read_result (value v)
+{
+  if (obj_tag(v) == Forward_tag)
+    return Field(v,0);
+  return v;
+}
+
+CAMLprim value caml_lazy_update_to_forcing (value v)
+{
+  tag_t tag;
+  value field0;
+
+  if (obj_update_tag (v, Lazy_tag, Forcing_tag)) {
```

Comment:
I think we should be making sure `v` is still a block before we pass it to `obj_update_tag`. In the case that `v` is not a block, this could lead to the same error we were witnessing previously. We should probably go over the other checks first before calling `obj_update_tag`.

Reply:
That's correct. Good catch.